tests/robustness/report/wal.go

```diff
@@ -0,0 +1,237 @@
+		if len(memberRequests) > len(persistedRequests) {
+			persistedRequests = memberRequests
+		}
+	}
+	return persistedRequests, nil
+}
+
+func requestsPersistedInWAL(lg *zap.Logger, dataDir string) ([]model.EtcdRequest, error) {
+	_, ents, err := ReadWAL(lg, dataDir)
+	if err != nil {
+		return nil, err
+	}
+	requests := make([]model.EtcdRequest, 0, len(ents))
+	for _, ent := range ents {
+		if ent.Type != raftpb.EntryNormal || len(ent.Data) == 0 {
+			continue
+		}
+		request, err := parseEntryNormal(ent)
+		if err != nil {
+			return nil, err
+		}
+		if request != nil {
+			requests = append(requests, *request)
+		}
+	}
+	return requests, nil
+}
+
+func ReadWAL(lg *zap.Logger, dataDir string) (state raftpb.HardState, ents []raftpb.Entry, err error) {
+	walDir := filepath.Join(dataDir, "member", "wal")
+	repaired := false
+	for {
+		w, err := wal.OpenForRead(lg, walDir, walpb.Snapshot{Index: 0})
+		if err != nil {
+			lg.Fatal("failed to open WAL", zap.Error(err))
+		}
+		_, state, ents, err = w.ReadAll()
+		w.Close()
+		if err != nil {
+			if errors.Is(err, wal.ErrSnapshotNotFound) {
+				return state, ents, nil
+			}
+			// we can only repair ErrUnexpectedEOF and we never repair twice.
+			if repaired || !errors.Is(err, io.ErrUnexpectedEOF) {
+				return state, nil, fmt.Errorf("failed to read WAL, cannot be repaired, err: %s", err)
+			}
+			if !wal.Repair(lg, walDir) {
+				return state, nil, fmt.Errorf("failed to repair WAL, err: %s", err)
+			}
+			lg.Info("repaired WAL", zap.Error(err))
+			repaired = true
+			continue
+		}
+		return state, ents, nil
+	}
+}
+
+func parseEntryNormal(ent raftpb.Entry) (*model.EtcdRequest, error) {
+	var raftReq pb.InternalRaftRequest
+	if err := raftReq.Unmarshal(ent.Data); err != nil {
+		var r pb.Request
+		isV2Entry := pbutil.MaybeUnmarshal(&r, ent.Data)
+		if !isV2Entry {
+			return nil, err
+		}
+		return nil, nil
+	}
+	switch {
+	case raftReq.Put != nil:
+		op := model.PutOptions{
+			Key:     string(raftReq.Put.Key),
+			Value:   model.ToValueOrHash(string(raftReq.Put.Value)),
+			LeaseID: raftReq.Put.Lease,
+		}
+		request := model.EtcdRequest{
+			Type: model.Txn,
+			Txn: &model.TxnRequest{
+				OperationsOnSuccess: []model.EtcdOperation{
+					{Type: model.PutOperation, Put: op},
+				},
+			},
+		}
+		return &request, nil
+	case raftReq.DeleteRange != nil:
+		op := model.DeleteOptions{Key: string(raftReq.DeleteRange.Key)}
+		request := model.EtcdRequest{
+			Type: model.Txn,
+			Txn: &model.TxnRequest{
+				OperationsOnSuccess: []model.EtcdOperation{
+					{Type: model.DeleteOperation, Delete: op},
+				},
+			},
+		}
+		return &request, nil
+	case raftReq.LeaseGrant != nil:
+		return &model.EtcdRequest{
+			Type:       model.LeaseGrant,
+			LeaseGrant: &model.LeaseGrantRequest{LeaseID: raftReq.LeaseGrant.ID},
+		}, nil
+	case raftReq.ClusterMemberAttrSet != nil:
+		return nil, nil
+	case raftReq.ClusterVersionSet != nil:
+		return nil, nil
+	case raftReq.Compaction != nil:
+		return nil, nil
```

Comment:
I was planning to start working on adding compaction support so that we can add it other tests like prevKV. I was going to do it through the recording client but I'm guessing if and when this merges, I can reuse the WAL itself.

Reply:
I have compaction implemented in draft PR https://github.com/etcd-io/etcd/pull/17680